Add Threshold-policy parameter support

chetanksmc opened the pull request:
Threshold policy parameter allows loadtime control of SE threshold values for different channels in different ways.
's' (static)     => driver keeps existing threshold value as-is
'd' (default)  => driver checks if SE threshold is less than Buf_len, if not, sets thresh as buf_len/2 and updates DMAC register
Int value       => driver will use custom threshold value provided by user for specific channel.

If a channel is not specified in threshold policy parameter, it will use 'default' policy.

Validated with Stamford with 4 channels in various combinations of above policy specifiers for different channels.

pvengat commented on switchtec_dma.c:
```diff
@@ -434,6 +454,57 @@ struct switchtec_dma_desc {
+					&thresh_kv_arr[i].thresh,
+					&thresh_kv_arr[i].ch_idx) == 2) {
+			thresh_kv_arr[i].policy = 'c'; /*User customized value*/
+		} else if (sscanf(ptr, "%c@%d",
```
But this would also match 6@5 (with '6' as a char)?

chetanksmc replied:
This is fine.

First if block with scanf(%d@%d), catches all integers. Only non-integers goto else if scanf(%c@%d) which catches characters.

pvengat commented on switchtec_dma.c:
```diff
@@ -3225,6 +3348,8 @@ static void switchtec_dma_remove(struct pci_dev *pdev)
 	dma_async_device_unregister(&swdma_dev->dma_dev);
 	put_device(swdma_dev->dma_dev.dev);
 
+	kfree(thresh_kv_arr);
```
On error case, thresh_kv_arr was freed. Need to have 'if' before freeing

chetanksmc replied:
Added if check before freeing the buffer wherever needed.

pvengat commented on switchtec_dma.c:
```diff
@@ -23,6 +23,26 @@ MODULE_VERSION(VERSION);
+		"                e.g. s@1,d@2,6@5\n"
+		"                s@1 => apply Static configuration value for channel 1 SE Threshold\n"
+		"                d@2 => setup Default configuration value for channel 1 SE Threshold(=SE_Buf_len/2)\n"
+		"                6@5 => set threshold value as 6 for channel 5channel 1 SE Threshold\n"
```
Typo?? Text says set threshold value as 6 for **channel 5channel 1**

chetanksmc replied:
Corrected typo.

pvengat commented on switchtec_dma.c:
```diff
@@ -434,6 +454,57 @@ struct switchtec_dma_desc {
+	ptr = thresh_policy;
+	do {
+		if(*ptr == ',') ptr++;
+		if (sscanf(ptr, "%d@%d",
```
This does not seem to be robust parsing & validation, can we use strtok_r?

chetanc10 replied:
strtok was complicating. I simplified using sscanf().
Though logic is updated to validate thresh channel/policy values properly.

chetanksmc replied:
strtok was complicating. But I updated logic to fully validate channel and policy values with kernel recommended string-manipulators.

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -23,6 +24,26 @@ MODULE_VERSION(VERSION);
+static char *thresh_policy = "";
+module_param(thresh_policy, charp, 0644);
+MODULE_PARM_DESC(thresh_policy, "<thresh-policy>\n"
+		"                Syntax: <thresh>@<channel-index>[,<threshx>@<channel-a>,...]\n"
```
The following pattern for the syntax might be better.
<se_threshold>@<channel>[,<se_threshold>@<channel>,...]

chetanksmc replied:
Updated

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -23,6 +24,26 @@ MODULE_VERSION(VERSION);
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Kelvin Cao");
 
+static char *thresh_policy = "";
```
se_thresh_policy?

chetanksmc replied:
Updated

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -23,6 +24,26 @@ MODULE_VERSION(VERSION);
+struct ch_thresh_s {
+    int ch_idx; /*Channel index*/
+    u32 thresh; /*Threshold value*/
+    char policy; /*Threshold policy 'd' or 's' or 'c'*/
```
Tab for indentation.

chetanksmc replied:
Done.

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -3180,6 +3334,10 @@ static int switchtec_dma_probe(struct pci_dev *pdev,
 
 	pci_set_master(pdev);
 
+	rc = setup_thresh_policy(pdev);
```
Processing argument and generating a global thresh_kv_arr array during probe complicates the handling and requires extra cleanup afterwards. Parsing the raw parameter string within switchtec_dma_chan_init() and handling the policy directly from it appears to be a cleaner approach.

chetanksmc replied:
Updated the code to have a simple logic to check for channel and get thresh policy info from se_thresh_policy during channel init as suggested.

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -1455,9 +1517,47 @@ static int switchtec_dma_chan_init(struct switchtec_dma_dev *swdma_dev, int i)
+			case 's': /*leave threshold as-is*/
+				thresh_ctrl = SE_THRESH_STATIC;
+				break;
+			case 'd': /*use default policy*/
```
This driver is not using many comments, especially in-line comments. Could you update to leave only necessary ones and not using in-line style?

chetanksmc replied:
Done. Removed extraneous inline comments and added a single explanatory comment for se-thresh-setting.

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -1455,9 +1517,47 @@ static int switchtec_dma_chan_init(struct switchtec_dma_dev *swdma_dev, int i)
+				orig_thresh = (valid_en_se >> SE_THRESH_SHIFT) &
+					SE_THRESH_MASK;
+				if (thresh != orig_thresh)
+					thresh_ctrl = SE_THRESH_NEW_VAL|SE_THRESH_SET_REG;
```
Spaces should be placed on both sides of the | operator.

chetanksmc replied:
Done

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -1455,9 +1517,47 @@ static int switchtec_dma_chan_init(struct switchtec_dma_dev *swdma_dev, int i)
+			return -1;
+		}
+		switch (policy) {
+			case 'c': /*Custom thresh value for this channel*/
```
The `case` lines should be aligned with the `switch` line.

chetanksmc replied:
Updated.

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -1455,9 +1517,47 @@ static int switchtec_dma_chan_init(struct switchtec_dma_dev *swdma_dev, int i)
+	if (thresh_ctrl & SE_THRESH_SET_REG) { /*update DMAC register*/
+		dev_info(dev, "Channel %d: Threshold orig=%u, %s=%u\n",
+				i, orig_thresh,
+				(thresh_ctrl & SE_THRESH_NEW_VAL) ? "new":"corrected",
```
Spaces should be placed on both sides of the ':'.

chetanksmc replied:
Done

kelvin-cao commented:
Looks good to me except some minor style issues. Please also squash the commits into one. Thanks.

chetanksmc committed:
Add Threshold-policy parameter support

Threshold-policy parameter is introduced to facilitate optimized
configuration tuning of individual DMA Channel thresholds while loading
the driver by avoiding the configuration-update-and-download cycles.
User can try different policies for channels in a comma-separated list.

Syntax: <se_threshold>@<channel-index>[,<se_thresholdx>@<channel-a>,...]

e.g. d@2,6@5,c@1
- d@2 means 'driver' policy i.e. ensure SE threshold for channel 2 is less
  than SE Buffer Length (if not, register value is set as 'se_buf_len/2')
- 6@5 means 'user' policy i.e. user can set 6 as channel 5 SE Threshold
  policy (se_buf_len/2)"
- c@1 means using previously configured value for channel 1 SE
  Threshold, the behavior for different cases is explained below.
  - First-time DMA-EP driver load:
    Driver uses threshold value set originally by Switch configuration tool.
  - Lateral driver loads with user-policies:
    2nd driver-load used 8@1, so 2nd driver-load sets 8 as Channel 1
    threshold. If 3rd driver-load used c@1, the threshold is set using
    previously configured value i.e. 8.

Note: Unspecified channels use driver policy

Signed-off-by: ChetaN KS <chetankumar.sanga@microchip.com>

chetanksmc commented:
> Looks good to me except some minor style issues. Please also squash the commits into one. Thanks.

Squashed to 1 commit.

lsgunth commented on switchtec_dma.c:
```diff
@@ -23,6 +23,17 @@ MODULE_VERSION(VERSION);
+		"                s@1 => apply Static configuration value for channel 1 SE Threshold\n"
+		"                d@2 => setup Default configuration value for channel 2 SE Threshold(=SE_Buf_len/2)\n"
+		"                6@5 => set threshold value as 6 for channel 5 SE Threshold\n"
+		"                NOTE: Channel index not specified in the list will be setup with Default configuration value(=SE_Buf_len/2)"
```
I feel like this documentation, and the commit message need expanding. What does SE stand for in this context?  What does the threshold actually do? What is a static configuration value? Why do we need a special way to specify the default? Couldn't we just leave it unset?

chetanksmc replied:
Thanks. I'll update the commit message with proper description of the feature added.

_> Why do we need a special way to specify the default? Couldn't we just leave it unset?_
I'll check on this and update.

chetanksmc replied:
Updated commit message to explain the feature.

lsgunth commented on switchtec_dma.c:
```diff
@@ -1455,9 +1514,46 @@ static int switchtec_dma_chan_init(struct switchtec_dma_dev *swdma_dev, int i)
+	}
+	if (thresh_ctrl & SE_THRESH_DEFAULT) {
+		thresh = (valid_en_se >> SE_THRESH_SHIFT) & SE_THRESH_MASK;
+		if (thresh >= se_buf_len) {
```
Probably should print a warning if the specified threshold cannot be set.

chetanksmc replied:
The dev_info() call at line 1549 already shows that thresh is changed to a 'corrected' value. A warning here feels redundant.

chetanksmc replied:
@lsgunth further comments?

lsgunth replied:
I don't know, up to you really. 

I find it a bit odd to always print the threshold and if the user doesn't get what they asked for they need to be very aware of what that message says.

chetanksmc replied:
Considering threshold print comes up during driver load only to notify changes in threshold, I'll keep this. Thanks.

lsgunth reviewed (commented):
All this seems a bit convoluted to me.

The threshold is configured once on initialization so the distinction between "new" and "corrected" is odd. If the value is important, just print what it is and allow the user to change it. Why is the old value that was set on reset important to anyone?

chetanksmc commented:
_> The threshold is configured once on initialization so the distinction between "new" and "corrected" is odd. If the value is important, just print what it is and allow the user to change it._
'new' => 'custom' threshold value given by user set in HW register as-is.
'corrected' => 'default' with correction that happens when HW register threshold value is not in range.
So the log with 'new' and 'corrected' distinction is added to indicate the above reasons of change.

_> Why is the old value that was set on reset important to anyone?_
The main purpose of se_thresh_policy parameter is to help users try some values in a set of test runs, so that user can easily select a optimal value per channel without having to going through a change-configuration-and-reset-switch cycle.
While reloading driver multiple times with different thresh policy settings, the 'old' value can be caused by a previous driver thresh setting to HW registers. The HW register will not reset unless a Switch reset happens. That is why we are also showing old values.

kelvin-cao commented:
> _> The threshold is configured once on initialization so the distinction between "new" and "corrected" is odd. If the value is important, just print what it is and allow the user to change it._ 'new' => 'custom' threshold value given by user set in HW register as-is. 'corrected' => 'default' with correction that happens when HW register threshold value is not in range. So the log with 'new' and 'corrected' distinction is added to indicate the above reasons of change.
> 
> _> Why is the old value that was set on reset important to anyone?_ The main purpose of se_thresh_policy parameter is to help users try some values in a set of test runs, so that user can easily select a optimal value per channel without having to going through a change-configuration-and-reset-switch cycle. While reloading driver multiple times with different thresh policy settings, the 'old' value can be caused by a previous driver thresh setting to HW registers. The HW register will not reset unless a Switch reset happens. That is why we are also showing old values.

Maybe we can take 'd' as driver policy instead of default policy as in this case driver overwrites the SE threshold value pre-configured via config file. Showing the original value gives user a hint what's happening during the driver initialization, and seems helpful.

chetanksmc commented:
@lsgunth @kelvin-cao 
Amended commit message with feature explanation.

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -23,6 +23,17 @@ MODULE_VERSION(VERSION);
+		"                Syntax: <se_threshold>@<channel-index>[,<se_thresholdx>@<channel-a>,...]\n"
+		"                e.g. s@1,d@2,6@5\n"
+		"                s@1 => apply 'static' configuration value for channel 1 SE Threshold\n"
+		"                d@2 => 'driver' policy - setup default configuration value for channel 2 SE Threshold(=SE_Buf_len/2)\n"
```
-> 'setup driver configuration'

chetanksmc replied:
Done

kelvin-cao commented on switchtec_dma.c:
```diff
@@ -1383,6 +1394,51 @@ static void switchtec_dma_free_chan_resources(struct dma_chan *chan)
+}
+
+#define SE_THRESH_STATIC  0x00
+#define SE_THRESH_DEFAULT 0x01
```
-> SE_THRESH_DRIVER

chetanksmc replied:
Done

lsgunth commented:
Reading the commit message I still don't know what his threshold policy actually does.  

I don't know what "static" means in this context.

One other question I had: Why do these thresholds have to be set with command line parameter? That may be inconvenient for a lot of cases. Is there a reason they can't be adjusted at runtime say through a sysfs parameter?

chetanksmc commented:
> Reading the commit message I still don't know what his threshold policy actually does.
> 
> I don't know what "static" means in this context.
> 
> One other question I had: Why do these thresholds have to be set with command line parameter? That may be inconvenient for a lot of cases. Is there a reason they can't be adjusted at runtime say through a sysfs parameter?

1. Threshold policy is an additional feature to help tuning of Channel SE thresholds only while loading driver, so that user may be able to tune channel thresholds without having to use Chiplink configuration regeneration and loading thus saving time.

2. Static means whatever Chiplink provided threshold configuration value for a Channel threshold will be used as-is i.e. a static value.

3. The threshold values for all channels are actually programmed during Channel-init sequence which in turn happens during DMA EP probe only. We don't allow for dynamic updates on Threshold values by design. So not using sysfs.

lsgunth commented:
> 1. Threshold policy is an additional feature to help tuning of Channel SE thresholds

Still haven't defined what SE is. And with out that, this is a circular definition. The threshold is the threshold. What is the threshold actually for?

> 2. Static means whatever Chiplink provided threshold configuration value for a Channel threshold will be used as-is i.e. a static value.

Ok, can we rename that? Static doesn't make any sense. I'd suggest  'config' to mean the value set in the switch's configuration data.


> 3. Static means whatever Chiplink provided threshold configuration value for a Channel threshold will be used as-is i.e. a static value.

I mean I still don't understand what this is adjusting, but it just appears to change the value of a hardware register. So when you say "we don't allow dynamic updates" is that because of the way these driver changes were written? Or is that actually inherent in the hardware? I don't see why the register couldn't be changed at any time or how the hardware would know that the value is being changed during "probe" or not.  It would be *substantially* more convenient for users to be able to just change a sysfs file to adjust a threshold register compared with having to adjust their kernel command line in an awkward way and reboot.

kelvin-cao commented:
> I don't see why the register couldn't be changed at any time or how the hardware would know that the value is being changed during "probe" or not.

Per hardware requirement, the SE (submission element, a basic DMA request structure) threshold must be configured before enabling a channel, in which case it seems allowing value changing during probe is appropriate, given this is more like a development oriented parameter where users wanted to fine tune the best parameters for a channel on their application setup, then use it at runtime without regular changes.

chetanksmc commented:
> > 1. Threshold policy is an additional feature to help tuning of Channel SE thresholds
> 
> Still haven't defined what SE is. And with out that, this is a circular definition. The threshold is the threshold. What is the threshold actually for?
> 
Okay, let me give the proper explanation of the feature being controlled here w.r.t DMA processing.
1. SE (Submission Element) is a DMA descriptor element, the driver creates SE buffers and make them available in a queue which the DMA EP processes and performs the necessary IO using the operation type, address, length and other info.
2. SE-Threshold is the maximum number of SE entries processed by DMA Controller (DMAC) at a time. 
3. SE-Buf-Len is the maximum number of SE entries that are prefetched by DMA Controller. Now if SE_THRESH value is recommended less than SE_BUF_LEN value to keep the DMAC operate continuously without running out of SE entries to process and waiting for next prefetch cycle, thus impacting performance. So the SE_THRESH needs to be set accordingly. This is actually available as configuration option in Chiplink tool where each channel has parameters defined to control SE_THRESH.

> > 2. Static means whatever Chiplink provided threshold configuration value for a Channel threshold will be used as-is i.e. a static value.
> 
> Ok, can we rename that? Static doesn't make any sense. I'd suggest 'config' to mean the value set in the switch's configuration data.
> 
Okay, if the name 'Static' doesn't relate to the underlying policy properly, we can use the name 'chiplink' instead, since the threshold value is originally setup by the Switch during FW init using the value from Chiplink generated configuration image per channel.
@kelvin-cao plz comment on name selection - 'config' or 'chiplink', which one?

> > 3. Static means whatever Chiplink provided threshold configuration value for a Channel threshold will be used as-is i.e. a static value.
> 
> I mean I still don't understand what this is adjusting, but it just appears to change the value of a hardware register. So when you say "we don't allow dynamic updates" is that because of the way these driver changes were written? Or is that actually inherent in the hardware? I don't see why the register couldn't be changed at any time or how the hardware would know that the value is being changed during "probe" or not. It would be _substantially_ more convenient for users to be able to just change a sysfs file to adjust a threshold register compared with having to adjust their kernel command line in an awkward way and reboot.

Yes, we are effectively changing HW register value. But the timing is important here. And the restriction is not from the driver. The DMAC HW has a sequence for channel specific registers to be set before a channel can be enabled. The SE_THRESH register one such register that must be set with Channel in disabled state. Since this is already happening in driver probe, we added threshold parameter updates as pre-channel-enablement code.
As we've already discussed this feature being for optional usage to ease fine-tuning of SE_THRESH parameter and not the final DMAC control approach, the users can only use this for initial testing and once an optimal SE_THRESH value is decided, users would then go to Chiplink and update the configuration value for relevant channel.
User can just reload driver (without having to reboot as it's all for testing phase only) with different thresh policy/values for different channels till user decides on final values for SE_THRESH.

Given this is an optional feature and not part of actual driver operation, we don't support sysfs parameter as that would need a major change in code to dynamically disable channel (this needs to clean up channel operational states properly), configure as per sysfs value and re-enable channel.
@kelvin-cao 
Maybe I can update the commit message to mark se_threshold_policy is only _an optional Host-Integration-Test-Phase feature_?

lsgunth commented:
>     2. SE-Threshold is the maximum number of SE entries processed by DMA Controller (DMAC) at a time.

I see. The language is a bit confusing as that doesn't sound like a threshold but the IO depth or something akin to that.  But I can appreciate if you need to use the term to match existing documentation. But please put a sentence like this in the commit message and parameter description.

> Yes, we are effectively changing HW register value. But the timing is important here. And the restriction is not from the driver. The DMAC HW has a sequence for channel specific registers to be set before a channel can be enabled. The SE_THRESH register one such register that must be set with Channel in disabled state. Since this is already happening in driver probe, we added threshold parameter updates as pre-channel-enablement code. As we've already discussed this feature being for optional usage to ease fine-tuning of SE_THRESH parameter and not the final DMAC control approach, the users can only use this for initial testing and once an optimal SE_THRESH value is decided, users would then go to Chiplink and update the configuration value for relevant channel. User can just reload driver (without having to reboot as it's all for testing phase only) with different thresh policy/values for different channels till user decides on final values for SE_THRESH.

Ok, I understand now. Thanks. That seems fine then.

kelvin-cao commented:
> Okay, if the name 'Static' doesn't relate to the underlying policy properly, we can use the name 'chiplink' instead, since the threshold value is originally setup by the Switch during FW init using the value from Chiplink generated configuration image per channel. @kelvin-cao plz comment on name selection - 'config' or 'chiplink', which one?

‘Config’ feels more appropriate, since Chiplink is one tool that modifies the configuration, but not the only way to do so.

> Given this is an optional feature and not part of actual driver operation, we don't support sysfs parameter as that would need a major change in code to dynamically disable channel (this needs to clean up channel operational states properly), configure as per sysfs value and re-enable channel. @kelvin-cao Maybe I can update the commit message to mark se_threshold_policy is only _an optional Host-Integration-Test-Phase feature_?

I guess it's not necessary, as it does provide an option for fine tuning channel.

Thanks.

chetanksmc commented:
@kelvin-cao 
Updated parameter description and naming as per latest review feedback. Updated commit message also to reflect the behavior of this parameter w.r.t the 3 policies.

Hope this is good to go now.

lsgunth reviewed (approved):
This looks good to me.

kelvin-cao reviewed (approved):
thanks.

lsgunth commented on switchtec_dma.c:
```diff
@@ -1455,9 +1514,45 @@ static int switchtec_dma_chan_init(struct switchtec_dma_dev *swdma_dev, int i)
+		}
+	}
+	if (thresh_ctrl & SE_THRESH_SET_REG) {
+		dev_info(dev, "Channel %d: Threshold orig=%u, %s=%u\n",
```
FYI: We received feedback from Vinod for this driver going upstream. One of the points was drivers should generally be silent:

https://lore.kernel.org/all/aUEnwQZQ383GgXMM@vaman/

We may need to change this to print only when there's an error or warning.

chetanksmc replied:
Sure, let's make it dev_dbg for thresh value prints. I looked at other dev_info()/pci_info() invocations. Are these okay?

1.  switchtec_dma_probe()
This pci_info() seems good as it shows driver's init proper, we can also add chan_cnt display here and remove chan_cnt pci_info() display mentioned in point 3.
```
struct switchtec_dma_dev *swdma_dev;
swdma_dev = pci_get_drvdata(pdev);
pci_info(pdev, "Switchtec DMA Channels Registered: %d\n", swdma_dev->chan_cnt);
```

2. switchtec_dma_create()
a. Either replace with pci_dbg() OR BETTER
b. Remove - original driver code differentiated PAX/fabric and other platforms. **But your upstream patch apparently removed PAX/fabric specific code for some reason, then there's no need to show identifying platform.**
```
pci_info(pdev, "Switchtec PSX/PFX DMA EP\n");
```

3. switchtec_dma_create()
If the change in point 1 is accepted, we can remove this line or else replace with pci_dbg():
```
pci_info(pdev, "Channel count: %d\n", chan_cnt);
```

4. switchtec_dma_create()
replace with pci_dbg()
```
pci_info(pdev, "Channel %d: %s\n", i++, dma_chan_name(chan));
```


@kelvin-cao
Logan's patch to upstream removed PAX/fabric specific code. May I know the reason?
